### PR TITLE
Run markdownlint on CI

### LINKED ISSUE #239: Investigate using a linter as part of CI process
[This project](https://github.com/DavidAnson/markdownlint) seems useful.

### PR DESCRIPTION
**Pull request by hugovk**
Fixes #239.

This installs and runs the [command-line interface](https://github.com/igorshubovych/markdownlint-cli) for [markdownlint](https://github.com/DavidAnson/markdownlint).  

Well, at the moment, there's a LOT of lint errors!

https://travis-ci.org/hugovk/mac-setup/builds/478022235#L754

643 to be exact:

```console
$ markdownlint . 2>&1 |  wc -l
     643
```

Here's a count by category:
```
   5 MD006/ul-start-left Consider starting bulleted lists at the beginning of the line
  13 MD007/ul-indent Unordered list indentation
   3 MD009/no-trailing-spaces Trailing spaces
  17 MD012/no-multiple-blanks Multiple consecutive blank lines
 209 MD013/line-length Line length
 144 MD014/commands-show-output Dollar signs used before commands without showing output
 127 MD022/blanks-around-headings/blanks-around-headers Headings should be surrounded by blank lines
  13 MD024/no-duplicate-heading/no-duplicate-header Multiple headings with the same content
   4 MD025/single-h1 Multiple top level headings in the same document
   1 MD028/no-blanks-blockquote Blank line inside blockquote
   9 MD029/ol-prefix Ordered list item prefix
  13 MD031/blanks-around-fences Fenced code blocks should be surrounded by blank lines
  41 MD032/blanks-around-lists Lists should be surrounded by blank lines
  30 MD033/no-inline-html Inline HTML
   1 MD034/no-bare-urls Bare URL used
   7 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading
   5 MD040/fenced-code-language Fenced code blocks should have a language specified
   1 MD042/no-empty-links No empty links
```

Are there any of these that should definitely be turned off right away? Or any that can be adjusted (eg. increase `line-length` a bit)?

I'd suggest tweaking the [config](https://github.com/igorshubovych/markdownlint-cli#configuration) a bit, fixing a few of the low-hanging fruit, then temporarily ignoring the others to get the CI to pass. Then the others can be tackled in batches one-by-one in their own PRs.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,7 +1,12 @@
 language: node_js
-cache: yarn
+cache:
+  - npm
+  - yarn
```
Comment (Kyslik) [inline .travis.yml]: Why are there two package managers used?

Comment (hugovk) [reply .travis.yml]: Good point, that'll teach me to copy and paste! Now only `yarn` is used.

**Comment by hugovk** [comment]
Ignoring "MD024/no-duplicate-heading/no-duplicate-header Multiple headings with the same content" because it complains about repeated headers like this, which I think are completely fine here:

## App1

### Installation

...

### Usage

...

## App2

### Installation

...

### Usage

...

**Comment by hugovk** [comment]
If using VS Code, I recommend the `markdownlint` (`davidanson.vscode-markdownlint`) extension, it shows you linting errors with a squiggly green underline and tooltips.

Down to 404 errors:
```
   5 MD006/ul-start-left Consider starting bulleted lists at the beginning of the line
  13 MD007/ul-indent Unordered list indentation
 208 MD013/line-length Line length
 144 MD014/commands-show-output Dollar signs used before commands without showing output
   2 MD028/no-blanks-blockquote Blank line inside blockquote
   9 MD029/ol-prefix Ordered list item prefix
  14 MD032/blanks-around-lists Lists should be surrounded by blank lines
   1 MD034/no-bare-urls Bare URL used
   7 MD036/no-emphasis-as-heading/no-emphasis-as-header Emphasis used instead of a heading
   1 MD042/no-empty-links No empty links
```

File: .travis.yml [hunk 2]
```diff
@@ -1,7 +1,10 @@
...
 node_js:
 - "8"
 install:
-- yarn install
-script: gitbook install && gitbook build
+  - yarn install
+  - yarn global add markdownlint-cli
+script:
+  - gitbook install && gitbook build
+  - markdownlint . --ignore node_modules
```
Comment (simeg) [inline .travis.yml]: Should we run the linter before building? Usually linting is quicker which would cause the CI to break faster.

**Comment by simeg** [comment]
@hugovk Awesome of you to pick this up! 😄I think your proposed strategy sounds good. If only the linter had a `--fix` flag 😛

File: .travis.yml (hunk 2 above)
Comment (hugovk) [reply .travis.yml]: Good idea, updated!

File: .travis.yml [hunk 3]
```diff
@@ -1,7 +1,10 @@
 language: node_js
-cache: yarn
+cache:  yarn
 node_js:
 - "8"
 install:
-- yarn install
-script: gitbook install && gitbook build
+  - yarn install
+  - yarn global add markdownlint-cli
```
Comment (simeg) [inline .travis.yml]: Could we move this to `package.json` under `devDependencies`? That way we would have all dependencies in one place.

**Comment by Kyslik** [comment]
I actually asked a few months ago about that `--fix` https://github.com/DavidAnson/vscode-markdownlint/issues/13#issuecomment-319719437 (for the VSC extension).

File: .markdownlint.json [hunk 4]
```diff
@@ -0,0 +1,4 @@
+{
+    "no-duplicate-heading": false,
+    "no-inline-html": false
```
Comment (Kyslik) [inline .markdownlint.json]: When I read this its kind counter-intuitive, double negative. 🙉 😆

File: .travis.yml (hunk 3 above)
Comment (hugovk) [reply .travis.yml]: Is it possible to globally install it that way? Or if not, how to run the command from a local install?

Comment (simeg) [reply .travis.yml]: We do it for gitbook so I think we could do it for this also.

Comment (hugovk) [reply .travis.yml]: Updated, it works!

File: .travis.yml [hunk 5]
```diff
@@ -1,7 +1,9 @@
...
+cache:  yarn
 node_js:
 - "8"
 install:
-- yarn install
-script: gitbook install && gitbook build
+  - yarn install
+script:
+  - markdownlint . --ignore node_modules
+  - gitbook install && gitbook build
```
Comment (simeg) [inline .travis.yml]: Looking at the logs Travis is not behaving as I expected, it doesn't fail fast even though the linting fails.

A pattern I think is nice is using Makefiles, that would give us some freedom to say "Only run X if Y succeeded" without having to write bash in this file. Then in this file we could just say "make ci" for script.

The other alternative would be to do `markdownlint . --ignore node_modules && gitbook install && gitbook build` but then I'd much prefer to just put it under make target.

See some discussions [here](https://github.com/travis-ci/travis-ci/issues/1066).

(Also gitbook install + build takes 60s and linting only 2s so having linting first makes sense)

Comment (simeg) [reply .travis.yml]: This change could be out of scope of this PR though, it's up to you.

Comment (hugovk) [reply .travis.yml]: One option from https://github.com/travis-ci/travis-ci/issues/1066: 

```yaml
  - set -e
  - command1
  - command2
  - set +e
```

Another is to have two build jobs, one for linting, one for build. But as you say, linting is so quick it's probably not worth the overhead of setting up the build job.

I'll do `command1 && command2` for this PR.

**Comment by hugovk** [comment]
Any comments about the remaining errors? Fix, tweak or ignore?

```
   5 MD006/ul-start-left Consider starting bulleted lists at the beginning of the line
  13 MD007/ul-indent Unordered list indentation
 208 MD013/line-length Line length
 144 MD014/commands-show-output Dollar signs used before commands without showing output
   2 MD028/no-blanks-blockquote Blank line inside blockquote
   9 MD029/ol-prefix Ordered list item prefix
  14 MD032/blanks-around-lists Lists should be surrounded by blank lines
   1 MD034/no-bare-urls Bare URL used
   1 MD042/no-empty-links No empty links
```

You can see the context from https://travis-ci.org/sb2nov/mac-setup/builds/478347674#L457.

If not, I'll disable those for the time being.

**Comment by hugovk** [comment]
If we allow failures then new PRs will introduce new linting errors and no-one will notice.

A build failure is a sure way to prevent new errors.

**Comment by Kyslik** [comment]
Maybe we can ignore MD013 (Line length) or no?

**Pull request merged**